File: packages/sdk-python/devbolt/validator.py
"""Configuration validator for DevBolt."""

import re
from typing import Any, Dict

from .types import (
    TARGETING_OPERATORS,
    FlagConfig,
    FlagsConfig,
    RolloutConfig,
    TargetingRule,
    ValidationError,
)

MAX_FLAG_NAME_LENGTH = 100
MAX_DESCRIPTION_LENGTH = 500
FLAG_NAME_REGEX = re.compile(r"^[a-z0-9_-]+$")


class ConfigValidator:
    """Validates feature flag configurations."""

    @staticmethod
    def validate(config: Any) -> None:
        """Validate entire flags configuration.

        Args:
            config: Configuration to validate

        Raises:
            ValidationError: If configuration is invalid
        """
        if not isinstance(config, dict):
            raise ValidationError("Config must be a dictionary")

        for flag_name, flag_config in config.items():
            ConfigValidator._validate_flag_name(flag_name)
            ConfigValidator._validate_flag_config(flag_name, flag_config)

    @staticmethod
    def _validate_flag_name(name: str) -> None:
        """Validate flag name format."""
        if not isinstance(name, str) or not name:
            raise ValidationError("Flag name must be a non-empty string", "flagName", name)

        if not FLAG_NAME_REGEX.match(name):
            raise ValidationError(
                f"Flag name '{name}' must contain only lowercase letters, numbers, underscores, and hyphens",
                "flagName",
                name,
            )

        if len(name) > MAX_FLAG_NAME_LENGTH:
            raise ValidationError(
                f"Flag name '{name}' exceeds maximum length of {MAX_FLAG_NAME_LENGTH}",
                "flagName",
                name,
            )

    @staticmethod
    def _validate_flag_config(flag_name: str, config: Any) -> None:
        """Validate flag configuration structure."""
        if not isinstance(config, dict):
            raise ValidationError(
                f"Flag '{flag_name}': config must be a dictionary", flag_name, config
            )

        # Validate required 'enabled' field
        if "enabled" not in config or not isinstance(config["enabled"], bool):
            raise ValidationError(
                f"Flag '{flag_name}': 'enabled' must be a boolean",
                f"{flag_name}.enabled",
                config.get("enabled"),
            )

        # Validate optional fields
        if "description" in config:
            ConfigValidator._validate_description(flag_name, config["description"])

        if "rollout" in config:
            ConfigValidator._validate_rollout(flag_name, config["rollout"])

        if "targeting" in config:
            ConfigValidator._validate_targeting(flag_name, config["targeting"])

        if "environments" in config:
            ConfigValidator._validate_environments(flag_name, config["environments"])

        if "metadata" in config:
            ConfigValidator._validate_metadata(flag_name, config["metadata"])

    @staticmethod
    def _validate_description(flag_name: str, description: Any) -> None:
        """Validate description field."""
        if not isinstance(description, str):
            raise ValidationError(
                f"Flag '{flag_name}': description must be a string",
                f"{flag_name}.description",
                description,
            )

        if len(description) > MAX_DESCRIPTION_LENGTH:
            raise ValidationError(
                f"Flag '{flag_name}': description exceeds maximum length of {MAX_DESCRIPTION_LENGTH}",
                f"{flag_name}.description",
                description,
            )

    @staticmethod
    def _validate_rollout(flag_name: str, rollout: Any) -> None:
        """Validate rollout configuration."""
        if not isinstance(rollout, dict):
            raise ValidationError(
                f"Flag '{flag_name}': rollout must be a dictionary",
                f"{flag_name}.rollout",
                rollout,
            )

        percentage = rollout.get("percentage")
        if not isinstance(percentage, (int, float)):
            raise ValidationError(
                f"Flag '{flag_name}': rollout.percentage must be a number",
                f"{flag_name}.rollout.percentage",
                percentage,
            )

        if not (0 <= percentage <= 100):
            raise ValidationError(
                f"Flag '{flag_name}': rollout.percentage must be between 0 and 100",
                f"{flag_name}.rollout.percentage",
                percentage,
            )

        if "seed" in rollout and not isinstance(rollout["seed"], str):
            raise ValidationError(
                f"Flag '{flag_name}': rollout.seed must be a string",
                f"{flag_name}.rollout.seed",
                rollout["seed"],
            )

    @staticmethod
    def _validate_targeting(flag_name: str, targeting: Any) -> None:
        """Validate targeting rules."""
        if not isinstance(targeting, list):
            raise ValidationError(
                f"Flag '{flag_name}': targeting must be a list",
                f"{flag_name}.targeting",
                targeting,
            )

        for index, rule in enumerate(targeting):
            ConfigValidator._validate_targeting_rule(flag_name, rule, index)

    @staticmethod
    def _validate_targeting_rule(flag_name: str, rule: Any, index: int) -> None:
        """Validate single targeting rule."""
        if not isinstance(rule, dict):
            raise ValidationError(
                f"Flag '{flag_name}': targeting rule {index} must be a dictionary",
                f"{flag_name}.targeting[{index}]",
                rule,
            )

        rule_key = f"{flag_name}.targeting[{index}]"

        # Validate attribute
        if (
            "attribute" not in rule
            or not isinstance(rule["attribute"], str)
            or not rule["attribute"]
        ):
            raise ValidationError(
                f"Flag '{flag_name}': targeting rule {index} attribute must be a non-empty string",
                f"{rule_key}.attribute",
                rule.get("attribute"),
            )

        # Validate operator
        if "operator" not in rule or rule["operator"] not in TARGETING_OPERATORS:
            raise ValidationError(
                f"Flag '{flag_name}': targeting rule {index} has invalid operator '{rule.get('operator')}'",
                f"{rule_key}.operator",
                rule.get("operator"),
            )

        operator = rule["operator"]

        # Validate value/values based on operator
        if operator in ("in", "not_in"):
            if "values" not in rule or not isinstance(rule["values"], list) or not rule["values"]:
                raise ValidationError(
                    f"Flag '{flag_name}': targeting rule {index} with operator '{operator}' requires non-empty 'values' list",
                    f"{rule_key}.values",
                    rule.get("values"),
                )
        else:
            if "value" not in rule:
                raise ValidationError(
                    f"Flag '{flag_name}': targeting rule {index} with operator '{operator}' requires 'value' field",
                    f"{rule_key}.value",
                    None,
                )

        # Validate enabled
        if "enabled" not in rule or not isinstance(rule["enabled"], bool):
            raise ValidationError(
                f"Flag '{flag_name}': targeting rule {index} 'enabled' must be a boolean",
                f"{rule_key}.enabled",
                rule.get("enabled"),
            )

        # Validate regex if operator is matches_regex
        if operator == "matches_regex":
            try:
                re.compile(str(rule["value"]))
            except re.error:
                raise ValidationError(
                    f"Flag '{flag_name}': targeting rule {index} has invalid regex pattern",
                    f"{rule_key}.value",
                    rule["value"],
                )

    @staticmethod
    def _validate_environments(flag_name: str, environments: Any) -> None:
        """Validate environments configuration."""
        if not isinstance(environments, dict):
            raise ValidationError(
                f"Flag '{flag_name}': environments must be a dictionary",
                f"{flag_name}.environments",
                environments,
            )

        for env, enabled in environments.items():
            if not isinstance(enabled, bool):
                raise ValidationError(
                    f"Flag '{flag_name}': environment '{env}' value must be a boolean",
                    f"{flag_name}.environments.{env}",
                    enabled,
                )

    @staticmethod
    def _validate_metadata(flag_name: str, metadata: Any) -> None:
        """Validate metadata."""
        if not isinstance(metadata, dict):
            raise ValidationError(
                f"Flag '{flag_name}': metadata must be a dictionary",
                f"{flag_name}.metadata",
                metadata,
            )

File: packages/core/src/validator.ts
import {
  FlagsConfig,
  FlagConfig,
  ValidationError,
  TARGETING_OPERATORS,
  type TargetingOperator,
} from "./types.js";

/**
 * Maximum allowed values
 */
const MAX_FLAG_NAME_LENGTH = 100;
const MAX_DESCRIPTION_LENGTH = 500;
const FLAG_NAME_REGEX = /^[a-z0-9_-]+$/;

/**
 * Validates feature flag configurations
 */
export class ConfigValidator {
  /**
   * Validate entire flags configuration
   */
  static validate(config: unknown): asserts config is FlagsConfig {
    if (typeof config !== "object" || config === null) {
      throw new ValidationError("Config must be an object");
    }

    if (Array.isArray(config)) {
      throw new ValidationError("Config must be an object, not an array");
    }

    const flags = config as Record<string, unknown>;
    const flagNames = Object.keys(flags);

    for (const flagName of flagNames) {
      this.validateFlagName(flagName);
      this.validateFlagConfig(flagName, flags[flagName]);
    }
  }

  /**
   * Validate flag name format
   */
  private static validateFlagName(name: string): void {
    if (typeof name !== "string" || name.length === 0) {
      throw new ValidationError(
        "Flag name must be a non-empty string",
        "flagName",
        name
      );
    }

    if (!FLAG_NAME_REGEX.test(name)) {
      throw new ValidationError(
        `Flag name "${name}" must contain only lowercase letters, numbers, underscores, and hyphens`,
        "flagName",
        name
      );
    }

    if (name.length > MAX_FLAG_NAME_LENGTH) {
      throw new ValidationError(
        `Flag name "${name}" exceeds maximum length of ${MAX_FLAG_NAME_LENGTH}`,
        "flagName",
        name
      );
    }
  }

  /**
   * Validate flag configuration structure
   */
  private static validateFlagConfig(
    flagName: string,
    config: unknown
  ): asserts config is FlagConfig {
    if (typeof config !== "object" || config === null) {
      throw new ValidationError(
        `Flag "${flagName}": config must be an object`,
        flagName,
        config
      );
    }

    const flag = config as Record<string, unknown>;

    // Validate required 'enabled' field
    if (typeof flag.enabled !== "boolean") {
      throw new ValidationError(
        `Flag "${flagName}": 'enabled' must be a boolean`,
        `${flagName}.enabled`,
        flag.enabled
      );
    }

    // Validate optional fields
    if (flag.description !== undefined) {
      this.validateDescription(flagName, flag.description);
    }

    if (flag.rollout !== undefined) {
      this.validateRollout(flagName, flag.rollout);
    }

    if (flag.targeting !== undefined) {
      this.validateTargeting(flagName, flag.targeting);
    }

    if (flag.environments !== undefined) {
      this.validateEnvironments(flagName, flag.environments);
    }

    if (flag.metadata !== undefined) {
      this.validateMetadata(flagName, flag.metadata);
    }
  }

  /**
   * Validate description field
   */
  private static validateDescription(
    flagName: string,
    description: unknown
  ): void {
    if (typeof description !== "string") {
      throw new ValidationError(
        `Flag "${flagName}": description must be a string`,
        `${flagName}.description`,
        description
      );
    }

    if (description.length > MAX_DESCRIPTION_LENGTH) {
      throw new ValidationError(
        `Flag "${flagName}": description exceeds maximum length of ${MAX_DESCRIPTION_LENGTH}`,
        `${flagName}.description`,
        description
      );
    }
  }

  /**
   * Validate rollout configuration
   */
  private static validateRollout(flagName: string, rollout: unknown): void {
    if (typeof rollout !== "object" || rollout === null) {
      throw new ValidationError(
        `Flag "${flagName}": rollout must be an object`,
        `${flagName}.rollout`,
        rollout
      );
    }

    const rolloutConfig = rollout as Record<string, unknown>;

    // Validate percentage
    if (typeof rolloutConfig.percentage !== "number") {
      throw new ValidationError(
        `Flag "${flagName}": rollout.percentage must be a number`,
        `${flagName}.rollout.percentage`,
        rolloutConfig.percentage
      );
    }

    if (!Number.isFinite(rolloutConfig.percentage)) {
      throw new ValidationError(
        `Flag "${flagName}": rollout.percentage must be finite`,
        `${flagName}.rollout.percentage`,
        rolloutConfig.percentage
      );
    }

    if (rolloutConfig.percentage < 0 || rolloutConfig.percentage > 100) {
      throw new ValidationError(
        `Flag "${flagName}": rollout.percentage must be between 0 and 100`,
        `${flagName}.rollout.percentage`,
        rolloutConfig.percentage
      );
    }

    // Validate optional seed
    if (
      rolloutConfig.seed !== undefined &&
      typeof rolloutConfig.seed !== "string"
    ) {
      throw new ValidationError(
        `Flag "${flagName}": rollout.seed must be a string`,
        `${flagName}.rollout.seed`,
        rolloutConfig.seed
      );
    }
  }

  /**
   * Validate targeting rules array
   */
  private static validateTargeting(flagName: string, targeting: unknown): void {
    if (!Array.isArray(targeting)) {
      throw new ValidationError(
        `Flag "${flagName}": targeting must be an array`,
        `${flagName}.targeting`,
        targeting
      );
    }

    targeting.forEach((rule, index) => {
      this.validateTargetingRule(flagName, rule, index);
    });
  }

  /**
   * Validate single targeting rule
   */
  private static validateTargetingRule(
    flagName: string,
    rule: unknown,
    index: number
  ): void {
    if (typeof rule !== "object" || rule === null) {
      throw new ValidationError(
        `Flag "${flagName}": targeting rule ${index} must be an object`,
        `${flagName}.targeting[${index}]`,
        rule
      );
    }

    const targetingRule = rule as Record<string, unknown>;
    const ruleKey = `${flagName}.targeting[${index}]`;

    // Validate attribute
    if (
      typeof targetingRule.attribute !== "string" ||
      targetingRule.attribute.length === 0
    ) {
      throw new ValidationError(
        `Flag "${flagName}": targeting rule ${index} attribute must be a non-empty string`,
        `${ruleKey}.attribute`,
        targetingRule.attribute
      );
    }

    // Validate operator
    if (
      !TARGETING_OPERATORS.includes(targetingRule.operator as TargetingOperator)
    ) {
      throw new ValidationError(
        `Flag "${flagName}": targeting rule ${index} has invalid operator "${targetingRule.operator}"`,
        `${ruleKey}.operator`,
        targetingRule.operator
      );
    }

    const operator = targetingRule.operator as TargetingOperator;

    // Validate value/values based on operator
    if (operator === "in" || operator === "not_in") {
      if (
        !Array.isArray(targetingRule.values) ||
        targetingRule.values.length === 0
      ) {
        throw new ValidationError(
          `Flag "${flagName}": targeting rule ${index} with operator "${operator}" requires non-empty 'values' array`,
          `${ruleKey}.values`,
          targetingRule.values
        );
      }

      // Validate each value in array
      for (const value of targetingRule.values) {
        if (
          typeof value !== "string" &&
          typeof value !== "number" &&
          typeof value !== "boolean"
        ) {
          throw new ValidationError(
            `Flag "${flagName}": targeting rule ${index} values must be string, number, or boolean`,
            `${ruleKey}.values`,
            value
          );
        }
      }
    } else {
      if (targetingRule.value === undefined) {
        throw new ValidationError(
          `Flag "${flagName}": targeting rule ${index} with operator "${operator}" requires 'value' field`,
          `${ruleKey}.value`,
          targetingRule.value
        );
      }

      // Validate value type
      const value = targetingRule.value;
      if (
        typeof value !== "string" &&
        typeof value !== "number" &&
        typeof value !== "boolean"
      ) {
        throw new ValidationError(
          `Flag "${flagName}": targeting rule ${index} value must be string, number, or boolean`,
          `${ruleKey}.value`,
          value
        );
      }
    }

    // Validate enabled field
    if (typeof targetingRule.enabled !== "boolean") {
      throw new ValidationError(
        `Flag "${flagName}": targeting rule ${index} 'enabled' must be a boolean`,
        `${ruleKey}.enabled`,
        targetingRule.enabled
      );
    }

    // Validate regex pattern if operator is matches_regex
    if (operator === "matches_regex") {
      try {
        new RegExp(String(targetingRule.value));
      } catch (error) {
        throw new ValidationError(
          `Flag "${flagName}": targeting rule ${index} has invalid regex pattern`,
          `${ruleKey}.value`,
          targetingRule.value
        );
      }
    }

    // Validate optional description
    if (targetingRule.description !== undefined) {
      if (typeof targetingRule.description !== "string") {
        throw new ValidationError(
          `Flag "${flagName}": targeting rule ${index} description must be a string`,
          `${ruleKey}.description`,
          targetingRule.description
        );
      }
    }
  }

  /**
   * Validate environments configuration
   */
  private static validateEnvironments(
    flagName: string,
    environments: unknown
  ): void {
    if (
      typeof environments !== "object" ||
      environments === null ||
      Array.isArray(environments)
    ) {
      throw new ValidationError(
        `Flag "${flagName}": environments must be an object`,
        `${flagName}.environments`,
        environments
      );
    }

    const envs = environments as Record<string, unknown>;

    for (const [env, enabled] of Object.entries(envs)) {
      if (typeof enabled !== "boolean") {
        throw new ValidationError(
          `Flag "${flagName}": environment "${env}" value must be a boolean`,
          `${flagName}.environments.${env}`,
          enabled
        );
      }
    }
  }

  /**
   * Validate metadata
   */
  private static validateMetadata(flagName: string, metadata: unknown): void {
    if (
      typeof metadata !== "object" ||
      metadata === null ||
      Array.isArray(metadata)
    ) {
      throw new ValidationError(
        `Flag "${flagName}": metadata must be an object`,
        `${flagName}.metadata`,
        metadata
      );
    }
  }
}
